### PR TITLE
[improve][broker] Add retry for start service unit state channel (ExtensibleLoadManagerImpl only)

### PR DESCRIPTION
**Pull request by Demogorgon314**
### Motivation

When the service unit state channel fails to start, it will throw an exception, and after the PR https://github.com/apache/pulsar/pull/22977 , it will fail forever, because the pulsar broker cannot catch the exception, and it won't have any retry mechanism

### Modifications

Add retry for start service unit state channel

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java [hunk 1]
```diff
@@ -879,6 +879,7 @@ private void preallocateBundle(String bundle, String broker) {
                                 k -> ConcurrentOpenHashMap.<String,
                                         ConcurrentOpenHashSet<String>>newBuilder()
                                         .build());
+        Set<String> myConcurrentSet = ConcurrentHashMap.newKeySet();
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:882]: Is this change required?

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java]: Removed, thanks

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 2]
```diff
@@ -401,10 +404,33 @@ public void start() throws PulsarServerException {
...
+                while (!Thread.currentThread().isInterrupted()) {
+                    try {
+                        brokerRegistry.register();
+                        this.serviceUnitStateChannel.start();
+                        break;
+                    } catch (Exception e) {
+                        log.warn("The broker:{} failed to start service unit state channel. Retrying {} th ...",
+                                pulsar.getBrokerId(), ++retry, e);
+                        try {
+                            TimeUnit.SECONDS.sleep(Math.min(retry * 10L, MAX_ROLE_CHANGE_RETRY_DELAY_IN_MILLIS));
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:417]: There's currently the org.apache.pulsar.common.util.Backoff class for doing exponential backoff with jitter. I think it would be better to use that solution instead of adding another way to do backoffs unless there's a specific reason for this logic.

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:417]: I want to set the max retry count, but since like the `Backoff` doesn't support it.

Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:417]: @Demogorgon314 Could you use the Backoff class for the delay calculation and keep the existing logic for max retry count?

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java]: Sure, it was updated.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 3]
```diff
@@ -498,8 +530,15 @@ private void failStarting(Exception ex) {
                 this.brokerRegistry, ex);
         if (this.brokerRegistry != null) {
             try {
-                brokerRegistry.close();
-            } catch (PulsarServerException e) {
+                brokerRegistry.unregister();
```
[the remark this answers is not in the document]
Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:533]: I use `unregister` here so we can `register` it back when retried.

[the remark this answers is not in the document]
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:533]: > It will be closed after `pulsar.close()`.

where?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 4]
```diff
@@ -401,10 +406,37 @@ public void start() throws PulsarServerException {
             this.serviceUnitStateChannel.listen(splitManager);
             this.leaderElectionService.start();
             pulsar.runWhenReadyForIncomingRequests(() -> {
-                try {
-                    this.serviceUnitStateChannel.start();
-                } catch (Exception e) {
-                    failStarting(e);
+                Backoff backoff = new BackoffBuilder()
+                        .setInitialTime(100, TimeUnit.MILLISECONDS)
+                        .setMax(conf.getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS)
```
Comment (RobertIndie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:411]: What's the reason to use `metadataStoreOperationTimeoutSeconds` as the max backoff time?

Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java]: I added a new field `STARTUP_TIMEOUT_SECONDS` as the timeout.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java [hunk 5]
```diff
@@ -401,10 +408,37 @@ public void start() throws PulsarServerException {
...
+                        break;
+                    } catch (Exception e) {
+                        log.warn("The broker:{} failed to start service unit state channel. Retrying {} th ...",
+                                pulsar.getBrokerId(), ++retry, e);
+                        try {
+                            Thread.sleep(backoff.next());
+                        } catch (InterruptedException ex) {
+                            log.warn("Interrupted while sleeping.");
+                            // preserve thread's interrupt status
+                            Thread.currentThread().interrupt();
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java:429]: Maybe we need to close the pulsar service if the thread is interrupted.

**Review by gaoran10 — APPROVED** [review]

**Comment by BewareMyPower** [comment]
Let me merge it since there is no new response from @lhotari. Please open a new PR if you found something wrong.
